### PR TITLE
ISSUE-80 WebSocket authentication via a ticket system

### PR DESCRIPTION
**Pull request by chibenwa**
TODO: Write a hell lot of tests.

Including:
 - [x] Contract for the TicketStore
 - [x] Tests with a MemoryTicketStore to exercise the clock logic of TicketManager
 - [x] Integration tests for the Auth Strategy including related routes (generation & revocation)
 - [x] Advertise the endpoints via a Session capability
Also...

 - [x] We need a CassandraTicketStore

Depends on https://github.com/apache/james-project/pull/334 (so build will fail)

### CODE REVIEW

**Review by Arsnael — COMMENTED** [review]
So far looks ok

Should we really use the ticket auth for normal http request as well? My guess it's a pain likely to distinguish the auth strategy between an http request and a websocket connection so I'm fine with it but just asking in case

**Comment by chibenwa** [comment]
```
Should we really use the ticket auth for normal http request as well? 
```

To be honest:
 - I do not see, security wise, some good reasons to restrict it.
 - Not having this distinction makes the code simpler.


```
My guess it's a pain likely to distinguish the auth strategy between an http request and a websocket connection so I'm fine with it but just asking in case
```

That's  why I did not do it...

**Comment by chibenwa** [comment]
IMO the Cassandra store can be contributed separately.

**Comment by chibenwa** [comment]
Woop woop!

**Comment by rouazana** [comment]
> ```
> 
> ```
> 
> 
> To be honest:
> 
> 
>     * I do not see, security wise, some good reasons to restrict it.
> 
> 
>     * Not having this distinction makes the code simpler.

I did not check the PR, but what's the behavior in case of double authentication?

**Comment by chibenwa** [comment]
I'll add a test.

**Comment by chibenwa** [comment]
What do you think should happen @rouazana ?

**Comment by rouazana** [comment]
> What do you think should happen @rouazana ?

I think that's a good reason to restrict it :)

Or else you can pick the same behavior than on JMAP(draft maybe?) authentication. If I remind well the idea was to disallow any double authentication.

**Comment by chibenwa** [comment]
Well the same `Authenticator` for the draft and rfc 8621 hence the behaviour is homogeneous.

We can duplicate the test here too no problem.

**Comment by chibenwa** [comment]
Needs https://github.com/apache/james-project/pull/336

File: openpaas-james/jmap/extensions/doc/ticketAuthentication.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
+# Ticket authentication
+
+We added support for WebSocket transport for JMAP. According to RFC-8887, the authentication is performed using 
+the `Authorization` header positioned upon the HTTP handshake when establishing the websocket connection.
+
+However, browser native APIs do not support setting additional headers upon the websocket HTTP handshake, defeating this
+strategy.
+
+To mitigate that, we cary the authentication payload in the URL, using a query parameter.
```
Comment (Arsnael) [inline openpaas-james/jmap/extensions/doc/ticketAuthentication.md:9]:
```suggestion
To mitigate that, we carry the authentication payload in the URL, using a query parameter.
```

File: openpaas-james/jmap/extensions/doc/ticketAuthentication.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+For security reasons, use a single use, short-lived, ip scoped, secondary ticket as an authentication mechanism.
+
+## Specification
+
+An authenticated client request a temporary token
```
Comment (Arsnael) [inline openpaas-james/jmap/extensions/doc/ticketAuthentication.md:15]:
```suggestion
An authenticated client requests a temporary token
```

File: openpaas-james/jmap/extensions/doc/ticketAuthentication.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+During the next 60s this ticket can be used a single time via query parameters as an authentication mechanism. Thus web-browser clients can use it to open a secure connection:
+
+```
+wss://james/jmap/js?ticket=UUID
+```
+
+The client discovers the ticket endpoints via an added capability in the session object:
+
+```
+"capabilities.com:linagora:params:jmap:ws:ticket": {
```
Comment (Arsnael) [inline openpaas-james/jmap/extensions/doc/ticketAuthentication.md:49]:
```suggestion
"capabilities:linagora:params:jmap:ws:ticket": {
```

**Comment by Arsnael** [comment]
```
[INFO] There are 2 errors reported by Checkstyle 8.29 with checkstyle.xml ruleset.
[ERROR] src/main/java/com/linagora/openpaas/james/app/MemoryServer.java:[12,8] (imports) UnusedImports: Unused import - org.apache.james.modules.protocols.JMAPServerModule.
[ERROR] src/main/java/com/linagora/openpaas/james/app/MemoryServer.java:[13,8] (imports) UnusedImports: Unused import - org.apache.james.modules.protocols.JmapEventBusModule.
```

File: openpaas-james/jmap/extensions-cassandra/pom.xml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            <groupId>${project.groupId}</groupId>
+            <artifactId>jmap-extensions</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>jmap-extensions</artifactId>
+            <type>test-jar</type>
+            <scope>test</scope>
+        </dependency>
+
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply openpaas-james/jmap/extensions-cassandra/pom.xml:24]: On purpose to separate `${project.groupId}` from `${project.groupId}`

File: openpaas-james/jmap/extensions-cassandra/pom.xml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            <artifactId>apache-james-backends-cassandra</artifactId>
+            <type>test-jar</type>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>${james.groupId}</groupId>
+            <artifactId>testing-base</artifactId>
+            <scope>test</scope>
+        </dependency>
+
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply openpaas-james/jmap/extensions-cassandra/pom.xml:40]: idem

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
```
[ERROR] com.linagora.openpaas.james.app.DistributedServerTest  Time elapsed: 43.729 s  <<< ERROR!
com.google.inject.ProvisionException: 
Unable to provision, see the following errors:

1) Error injecting constructor, com.datastax.driver.core.exceptions.InvalidQueryException: unconfigured table ticket
  at com.linagora.openpaas.james.jmap.ticket.CassandraTicketStore.<init>(CassandraTicketStore.scala:50)
  at com.linagora.openpaas.james.jmap.ticket.CassandraTicketStoreModule.configure(CassandraTicketStore.scala:22) (via modules: com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$OverrideModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$CombinedModule -> com.google.inject.util.Modules$OverrideModule -> com.linagora.openpaas.james.jmap.ticket.CassandraTicketStoreModule)
  while locating com.linagora.openpaas.james.jmap.ticket.CassandraTicketStore
  while locating com.linagora.openpaas.james.jmap.ticket.TicketStore
    for the 2nd parameter of com.linagora.openpaas.james.jmap.ticket.TicketManager.<init>(TicketManager.scala:35)
  while locating com.linagora.openpaas.james.jmap.ticket.TicketManager
    for the 1st parameter of com.linagora.openpaas.james.jmap.ticket.TicketAuthenticationStrategy.<init>(TicketAuthenticationStrategy.scala:19)
  while locating com.linagora.openpaas.james.jmap.ticket.TicketAuthenticationStrategy
  while locating org.apache.james.jmap.http.AuthenticationStrategy annotated with @com.google.inject.internal.Element(setName=,uniqueId=170, type=MULTIBINDER, keyType=)
  while locating java.util.Set<org.apache.james.jmap.http.AuthenticationStrategy>
[...]
```

**Comment by chibenwa** [comment]
Looks related :-P

**Review by Arsnael — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

File: openpaas-james/jmap/extensions/src/test/scala/com/linagora/openpaas/james/jmap/ticket/TicketManagerTest.scala [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  private val initialDate = ZonedDateTime.parse("2010-10-30T15:12:00Z")
+  private val username = Username.of("bob")
+  private val address = InetAddress.getByName("127.0.0.1")
+  var clock: UpdatableTickingClock = null
+  var testee: TicketManager = null
+
+  @BeforeEach
+  def setUp(): Unit = {
+    clock = new UpdatableTickingClock(initialDate.toInstant)
+    testee =  new TicketManager(clock, new MemoryTicketStore())
```
Comment (LanKhuat) [inline openpaas-james/jmap/extensions/src/test/scala/com/linagora/openpaas/james/jmap/ticket/TicketManagerTest.scala:21]: extra space

**Review by LanKhuat — APPROVED** [review]

**Comment by Arsnael** [comment]
Need rebase

**Review by Arsnael — APPROVED** [review]

**Review by Arsnael — COMMENTED** [review]
Submodule update? I guess it's the latest master on apache james that has not been merged yet though on master in openpaas-james

**Comment by chibenwa** [comment]
That's the latest master.

**Pull request merged**